File: ga/glm4_result_2024_6/result/tree_glm/28_tree.c
use std::ptr;

type ElementType = i32;

struct TreeNode {
    element: ElementType,
    left: Option<Box<TreeNode>>,
    right: Option<Box<TreeNode>>,
}

type SearchTree = Option<Box<TreeNode>>;

fn make_empty(tree: SearchTree) -> SearchTree {
    if let Some(mut node) = tree {
        node.left = make_empty(node.left);
        node.right = make_empty(node.right);
        None
    } else {
        None
    }
}

fn find(x: ElementType, tree: &SearchTree) -> Option<&TreeNode> {
    let mut current = tree.as_ref()?;
    while current.is_some() {
        if x < current.as_ref().unwrap().element {
            current = current.as_ref().unwrap().left.as_ref();
        } else if x > current.as_ref().unwrap().element {
            current = current.as_ref().unwrap().right.as_ref();
        } else {
            return current;
        }
    }
    None
}

fn find_min(tree: &SearchTree) -> Option<&TreeNode> {
    let mut current = tree.as_ref()?;
    while let Some(node) = current {
        if node.left.is_none() {
            return Some(node);
        }
        current = node.left.as_ref();
    }
    None
}

fn find_max(tree: &SearchTree) -> Option<&TreeNode> {
    let mut current = tree.as_ref()?;
    while let Some(node) = current {
        if node.right.is_none() {
            return Some(node);
        }
        current = node.right.as_ref();
    }
    None
}

fn insert(x: ElementType, tree: SearchTree) -> SearchTree {
    if let None = tree {
        return Some(Box::new(TreeNode {
            element: x,
            left: None,
            right: None,
        }));
    }

    let mut current = tree;
    loop {
        match current.as_mut() {
            Some(node) => {
                if x < node.element {
                    current = node.left.insert(x);
                } else if x > node.element {
                    current = node.right.insert(x);
                } else {
                    return current;
                }
            }
            None => return Some(Box::new(TreeNode {
                element: x,
                left: None,
                right: None,
            })),
        }
    }
}

fn delete(x: ElementType, tree: SearchTree) -> SearchTree {
    if let Some(mut node) = tree {
        if x < node.element {
            node.left = delete(x, node.left);
        } else if x > node.element {
            node.right = delete(x, node.right);
        } else {
            if let Some(mut right) = node.right.take() {
                node.element = find_min(&right).unwrap().element;
                node.right = delete(node.element, right);
            } else {
                return node.left;
            }
        }
        Some(node)
    } else {
        None
    }
}

fn retrieve(p: &TreeNode) -> ElementType {
    p.element
}

fn main() {
    let mut t: SearchTree = None;
    for i in 0..50 {
        t = insert(i, t);
    }
    for i in 0..50 {
        if let None = find(i, &t) {
            println!("Error at {}", i);
        }
    }
    for i in (0..50).step_by(2) {
        t = delete(i, t);
    }
    for i in (1..50).step_by(2) {
        if let None = find(i, &t) {
            println!("Error at {}", i);
        }
    }
    for i in (0..50).step_by(2) {
        if let Some(_) = find(i, &t) {
            println!("Error at {}", i);
        }
    }
    println!("Min is {}, Max is {}", retrieve(find_min(&t).unwrap()), retrieve(find_max(&t).unwrap()));
}

File: ga/4omini_result_2024_7/result/tree_mini/28_tree.rs
use std::alloc::{alloc, dealloc, Layout};
use std::ptr;

type ElementType = i32;

struct TreeNode {
    element: ElementType,
    left: Option<Box<TreeNode>>,
    right: Option<Box<TreeNode>>,
}

struct SearchTree {
    root: Option<Box<TreeNode>>,
}

impl SearchTree {
    fn new() -> Self {
        SearchTree { root: None }
    }

    fn make_empty(&mut self) {
        self.root = None;
    }

    fn find(&self, x: ElementType) -> Option<&TreeNode> {
        self.find_node(x, &self.root)
    }

    fn find_node<'a>(&'a self, x: ElementType, node: &'a Option<Box<TreeNode>>) -> Option<&'a TreeNode> {
        match node {
            Some(n) => {
                if x < n.element {
                    self.find_node(x, &n.left)
                } else if x > n.element {
                    self.find_node(x, &n.right)
                } else {
                    Some(n)
                }
            }
            None => None,
        }
    }

    fn find_min(&self) -> Option<&TreeNode> {
        self.find_min_node(&self.root)
    }

    fn find_min_node<'a>(&'a self, node: &'a Option<Box<TreeNode>>) -> Option<&'a TreeNode> {
        match node {
            Some(n) => {
                if n.left.is_none() {
                    Some(n)
                } else {
                    self.find_min_node(&n.left)
                }
            }
            None => None,
        }
    }

    fn find_max(&self) -> Option<&TreeNode> {
        self.find_max_node(&self.root)
    }

    fn find_max_node<'a>(&'a self, node: &'a Option<Box<TreeNode>>) -> Option<&'a TreeNode> {
        let mut current = node;
        while let Some(n) = current {
            if n.right.is_none() {
                return Some(n);
            }
            current = &n.right;
        }
        None
    }

    fn insert(&mut self, x: ElementType) {
        self.root = Self::insert_node(x, self.root.take());
    }

    fn insert_node(node: ElementType, tree: Option<Box<TreeNode>>) -> Option<Box<TreeNode>> {
        match tree {
            None => Some(Box::new(TreeNode {
                element: node,
                left: None,
                right: None,
            })),
            Some(mut n) => {
                if node < n.element {
                    n.left = Self::insert_node(node, n.left);
                } else if node > n.element {
                    n.right = Self::insert_node(node, n.right);
                }
                Some(n)
            }
        }
    }

    fn delete(&mut self, x: ElementType) {
        self.root = Self::delete_node(x, self.root.take());
    }

    fn delete_node(tree: ElementType, node: Option<Box<TreeNode>>) -> Option<Box<TreeNode>> {
        match node {
            None => {
                eprintln!("Element not found");
                None
            }
            Some(mut n) => {
                if tree < n.element {
                    n.left = Self::delete_node(tree, n.left);
                } else if tree > n.element {
                    n.right = Self::delete_node(tree, n.right);
                } else {
                    match (n.left.take(), n.right.take()) {
                        (None, None) => return None,
                        (Some(left), None) => return Some(left),
                        (None, Some(right)) => return Some(right),
                        (Some(left), Some(right)) => {
                            let min_right = Self::find_min_node(&Some(right.clone()));
                            n.element = min_right.unwrap().element; // Replace with min of right subtree
                            n.right = Self::delete_node(n.element, Some(right));
                        }
                    }
                }
                Some(n)
            }
        }
    }

    fn retrieve(&self, node: &Option<Box<TreeNode>>) -> Option<ElementType> {
        node.as_ref().map(|n| n.element)
    }
}

fn main() {
    let mut tree = SearchTree::new();

    let mut j = 0;
    for i in 0..50 {
        j = (j + 7) % 50;
        tree.insert(j);
    }

    for i in 0..50 {
        match tree.find(i) {
            Some(node) if node.element != i => {
                println!("Error at {}", i);
            }
            None => {
                println!("Error at {}", i);
            }
            _ => {}
        }
    }

    for i in (0..50).step_by(2) {
        tree.delete(i);
    }

    for i in (1..50).step_by(2) {
        match tree.find(i) {
            Some(node) if node.element != i => {
                println!("Error at {}", i);
            }
            None => {
                println!("Error at {}", i);
            }
            _ => {}
        }
    }

    for i in (0..50).step_by(2) {
        if tree.find(i).is_some() {
            println!("Error at {}", i);
        }
    }

    if let (Some(min), Some(max)) = (tree.find_min(), tree.find_max()) {
        println!("Min is {}, Max is {}", min.element, max.element);
    }
}